File: xim/resource/ZoneMeshSection.kt
package xim.resource

import xim.math.Vector2f
import xim.math.Vector3f
import xim.poc.gl.*
import xim.resource.table.ZoneDecrypt
import kotlin.math.abs

class ZoneMeshSection(private val sectionHeader: SectionHeader) : ResourceParser {

    private lateinit var name: String
    private val buffers = ArrayList<MeshBuffer>()

    private var boundingBox0: BoundingBox? = null
    private var boundingBox1: BoundingBox? = null

    override fun getResource(byteReader: ByteReader): ParserResult {
        ZoneDecrypt.decryptZoneMesh(sectionHeader, byteReader)
        read(byteReader)
        val resource = ZoneMeshResource(sectionHeader.sectionId, buffers, name, boundingBox0, boundingBox1)
        return ParserResult(ParserEntry(resource, name))
    }

    private fun read(byteReader: ByteReader) {
        byteReader.offsetFromDataStart(sectionHeader)

        val decryptInfo = byteReader.next32() // DecodeInfo - not needed here
        val keyConfigData = byteReader.next32() // Decode key + section specific params
        val unkString = byteReader.nextString(0x8) // Author name?

        val configData = keyConfigData and 0x000000FF

        val meshTypeFlag = configData and 0x1
        val meshType = if (meshTypeFlag == 0) { MeshType.TriMesh } else { MeshType.TriStrip }
        val vertexBlendEnabled = (configData and 0x2) != 0

        name = byteReader.nextString(0x10).trimEnd()

        val defStart = byteReader.position

        // Section 1
        val meshCount0 = byteReader.next32()
        boundingBox0 = BoundingBox.read(byteReader)

        val section1DataStart = defStart + byteReader.next32()

        // "hit"-type models, only have a bounding box & no data
        if (meshCount0 == 0) {
            if (byteReader.position != sectionHeader.sectionStartPosition + sectionHeader.sectionSize) {
                throw IllegalStateException("Expected to be at the end")
            }

            return
        }

        // Section 2
        val meshCount1 = byteReader.next32()
        boundingBox1 = BoundingBox.read(byteReader)

        // Not sure what the last 4 bytes are
        val unknown = byteReader.next32()
        if (byteReader.position != section1DataStart) {
            throw IllegalStateException("More defn data than expected")
        }

        for (i in 0 until meshCount1) {
            val meshBuffer = parseMesh(byteReader, name, vertexBlendEnabled, meshType)
            buffers.add(meshBuffer)
            byteReader.align0x04()
        }
    }

    private fun parseMesh(byteReader: ByteReader, name: String, vertexBlendEnabled: Boolean, meshType: MeshType): MeshBuffer {
        val textureName = byteReader.nextString(0x10)
        val numVerts = byteReader.next16()
        val verts = ArrayList<Vertex>(numVerts)

        val flags = byteReader.next16()
        val blendEnabled = (flags and 0x8000) != 0
        val backFaceCulling = (flags and 0x2000) == 0 // by default, culling is enabled to CCW & setting this flag will disable it

        for(i in 0 until numVerts) {
            if (vertexBlendEnabled) {
                verts.add(Vertex(
                    byteReader.nextVector3f(),
                    byteReader.nextVector3f(),
                    byteReader.nextVector3f(),
                    ByteColor(b = byteReader.next8(), g = byteReader.next8(), r = byteReader.next8(), a = byteReader.next8()),
                    byteReader.nextFloat(), byteReader.nextFloat()
                ))
            } else {
                verts.add(Vertex(
                    byteReader.nextVector3f(),
                    Vector3f(),
                    byteReader.nextVector3f(),
                    ByteColor(b = byteReader.next8(), g = byteReader.next8(), r = byteReader.next8(), a = byteReader.next8()),
                    byteReader.nextFloat(), byteReader.nextFloat()
                ))
            }
        }

        val numIndicies = byteReader.next16()
        val unk1 = byteReader.next16()
        val indexedVertices = ArrayList<Vertex>(numIndicies)

        for (i in 0 until numIndicies) {
            val idx = byteReader.next16()
            indexedVertices += verts[idx]
        }

        val glBufferBuilder = GlBufferBuilder(numIndicies)
        for (i in 0 until numIndicies) {
            val tangent = computeTangent(i, indexedVertices, meshType)
            glBufferBuilder.appendZoneMeshVertex(indexedVertices[i], tangent)
        }

        val bufferId = glBufferBuilder.build()

        val discardThreshold = if (name.startsWith("_")) { 0.375f } else { null }

        val zBias = when (blendEnabled) {
            true -> ZBiasLevel.High
            false -> ZBiasLevel.Normal
        }

        return MeshBuffer(
            numVertices = numIndicies,
            meshType = meshType,
            textureStage0 = TextureLink.of(textureName, sectionHeader.localDir),
            renderState = RenderState(blendEnabled = blendEnabled, zBias = zBias, discardThreshold = discardThreshold, useBackFaceCulling = backFaceCulling),
            blendVertexPosition = vertexBlendEnabled,
            glBuffer = bufferId
        )
    }

    private fun computeTangent(i: Int, verts: List<Vertex>, type: MeshType): Vector3f {
        if (i <= 2) { return computeTangent(verts[0], verts[1], verts[2]) }

        return when (type) {
            MeshType.TriMesh -> {
                val i0 = i - (i % 3)
                computeTangent(verts[i0], verts[i0+1], verts[i0+2])
            }
            MeshType.TriStrip -> {
                computeTangent(verts[i-2], verts[i-1], verts[i])
            }
            MeshType.TriFan -> {
                computeTangent(verts[0], verts[i-1], verts[i])
            }
        }
    }

    private fun computeTangent(v0: Vertex, v1: Vertex, v2: Vertex): Vector3f {
        val edge1 = v1.p0 - v0.p0
        val edge2 = v2.p0 - v0.p0

        val dUV1 = Vector2f(v1.texCoordU, v1.texCoordV) - Vector2f(v0.texCoordU, v0.texCoordV)
        val dUV2 = Vector2f(v2.texCoordU, v2.texCoordV) - Vector2f(v0.texCoordU, v0.texCoordV)

        val denom = dUV1.x * dUV2.y - dUV2.x * dUV1.y
        if (abs(denom) < 1e-5) { return Vector3f(1f, 1f, 1f).normalizeInPlace() }

        val f = 1.0f / denom
        return Vector3f(
            x = f * (dUV2.y * edge1.x - dUV1.y * edge2.x),
            y = f * (dUV2.y * edge1.y - dUV1.y * edge2.y),
            z = f * (dUV2.y * edge1.z - dUV1.y * edge2.z),
        )
    }

    data class BoundingBox(
        val p0: Vector3f,
        val p1: Vector3f,
    ) {

        companion object {
            fun read(byteReader: ByteReader): BoundingBox {
                val x0 = byteReader.nextFloat()
                val x1 = byteReader.nextFloat()

                val y0 = byteReader.nextFloat()
                val y1 = byteReader.nextFloat()

                val z0 = byteReader.nextFloat()
                val z1 = byteReader.nextFloat()

                return BoundingBox(Vector3f(x0, y0, z0), Vector3f(x1, y1, z1))
            }
        }
    }

    class Vertex(
        val p0: Vector3f,
        val p1: Vector3f,
        val n0: Vector3f,
        val colorMask: ByteColor,
        val texCoordU: Float,
        val texCoordV: Float,
    )

}

File: lib/resource/zoneMeshSection.ts
/**
 * Zone mesh section parser (section type 0x2E).
 *
 * Ported from xim/resource/ZoneMeshSection.kt.
 * Parses encrypted zone mesh data into ZoneMeshResource.
 */

import type { ByteReader, Vector3 } from './byteReader'
import type { SectionHeader } from './datParser'
import { ParserEntry, ParserResult, type ResourceParser } from './datParser'
import { DatId, type MeshRenderState } from './datResource'
import { TextureLink } from './textureLink'
import {
  ZoneMeshResource,
  type ZoneMeshBoundingBox,
  type ZoneMeshBuffer,
  type ZoneMeshVertex,
} from './zoneResource'
import { decryptZoneMesh } from './table/zoneDecrypt'

export class ZoneMeshSection implements ResourceParser {
  private readonly sectionHeader: SectionHeader

  constructor(sectionHeader: SectionHeader) {
    this.sectionHeader = sectionHeader
  }

  getResource(byteReader: ByteReader): ParserResult {
    decryptZoneMesh(this.sectionHeader, byteReader)
    const { name, buffers, boundingBox0, boundingBox1 } = this.read(byteReader)
    const resource = new ZoneMeshResource(
      this.sectionHeader.sectionId,
      name,
      buffers,
      boundingBox0,
      boundingBox1,
    )
    return new ParserResult(new ParserEntry(resource, name))
  }

  private read(byteReader: ByteReader): {
    name: string
    buffers: ZoneMeshBuffer[]
    boundingBox0: ZoneMeshBoundingBox | null
    boundingBox1: ZoneMeshBoundingBox | null
  } {
    byteReader.offsetFromDataStart(this.sectionHeader)

    const _decryptInfo = byteReader.next32()
    const keyConfigData = byteReader.next32()
    const _unkString = byteReader.nextString(0x8)

    const configData = keyConfigData & 0x000000ff
    const meshTypeFlag = configData & 0x1
    const meshType: 'TriMesh' | 'TriStrip' = meshTypeFlag === 0 ? 'TriMesh' : 'TriStrip'
    const vertexBlendEnabled = (configData & 0x2) !== 0

    const name = byteReader.nextString(0x10).replace(/\0+$/, '').trimEnd()

    const defStart = byteReader.position

    // Section 1
    const meshCount0 = byteReader.next32()
    const boundingBox0 = readBoundingBox(byteReader)
    const section1DataStart = defStart + byteReader.next32()

    // "hit"-type models have only a bounding box with no data
    if (meshCount0 === 0) {
      return { name, buffers: [], boundingBox0, boundingBox1: null }
    }

    // Section 2
    const meshCount1 = byteReader.next32()
    const boundingBox1 = readBoundingBox(byteReader)
    const _unknown = byteReader.next32()

    if (byteReader.position !== section1DataStart) {
      console.warn(`[ZoneMesh] More definition data than expected`)
    }

    const buffers: ZoneMeshBuffer[] = []
    for (let i = 0; i < meshCount1; i++) {
      buffers.push(this.parseMesh(byteReader, name, vertexBlendEnabled, meshType))
      byteReader.align0x04()
    }

    return { name, buffers, boundingBox0, boundingBox1 }
  }

  private parseMesh(
    byteReader: ByteReader,
    name: string,
    vertexBlendEnabled: boolean,
    meshType: 'TriMesh' | 'TriStrip',
  ): ZoneMeshBuffer {
    const textureName = byteReader.nextString(0x10)
    const numVerts = byteReader.next16()
    const flags = byteReader.next16()
    const blendEnabled = (flags & 0x8000) !== 0
    const backFaceCulling = (flags & 0x2000) === 0

    const verts: ZoneMeshVertex[] = []
    for (let i = 0; i < numVerts; i++) {
      if (vertexBlendEnabled) {
        verts.push({
          p0: byteReader.nextVector3f(),
          p1: byteReader.nextVector3f(),
          normal: byteReader.nextVector3f(),
          color: byteReader.nextBGRA(),
          u: byteReader.nextFloat(),
          v: byteReader.nextFloat(),
        })
      } else {
        verts.push({
          p0: byteReader.nextVector3f(),
          p1: { x: 0, y: 0, z: 0 },
          normal: byteReader.nextVector3f(),
          color: byteReader.nextBGRA(),
          u: byteReader.nextFloat(),
          v: byteReader.nextFloat(),
        })
      }
    }

    const numIndices = byteReader.next16()
    const _unk1 = byteReader.next16()

    const indices: number[] = []
    const indexedVertices: ZoneMeshVertex[] = []
    for (let i = 0; i < numIndices; i++) {
      const idx = byteReader.next16()
      indices.push(idx)
      indexedVertices.push(verts[idx]!)
    }

    // Compute tangent vectors per-triangle for bump mapping
    const tangents: Vector3[] = []
    for (let i = 0; i < numIndices; i++) {
      tangents.push(computeTangent(i, indexedVertices, meshType))
    }

    const discardThreshold = name.startsWith('_') ? 0.375 : null

    const zBias = blendEnabled ? 5 : 0

    const renderState: MeshRenderState = {
      blendEnabled,
      discardThreshold,
      depthMask: !blendEnabled,
      useBackFaceCulling: backFaceCulling,
      zBias,
    }

    return {
      numVertices: numIndices,
      meshType,
      textureName,
      textureLink: TextureLink.of(textureName, this.sectionHeader.localDir),
      vertices: indexedVertices,
      indices,
      tangents,
      renderState,
      vertexBlendEnabled,
    }
  }
}

function readBoundingBox(byteReader: ByteReader): ZoneMeshBoundingBox {
  const x0 = byteReader.nextFloat()
  const x1 = byteReader.nextFloat()
  const y0 = byteReader.nextFloat()
  const y1 = byteReader.nextFloat()
  const z0 = byteReader.nextFloat()
  const z1 = byteReader.nextFloat()
  return {
    p0: { x: x0, y: y0, z: z0 },
    p1: { x: x1, y: y1, z: z1 },
  }
}

function computeTangent(i: number, verts: readonly ZoneMeshVertex[], type: 'TriMesh' | 'TriStrip'): Vector3 {
  if (i <= 2) return computeTangentFromTriangle(verts[0]!, verts[1]!, verts[2]!)

  if (type === 'TriMesh') {
    const i0 = i - (i % 3)
    return computeTangentFromTriangle(verts[i0]!, verts[i0 + 1]!, verts[i0 + 2]!)
  }

  // TriStrip
  return computeTangentFromTriangle(verts[i - 2]!, verts[i - 1]!, verts[i]!)
}

function computeTangentFromTriangle(v0: ZoneMeshVertex, v1: ZoneMeshVertex, v2: ZoneMeshVertex): Vector3 {
  const edge1x = v1.p0.x - v0.p0.x
  const edge1y = v1.p0.y - v0.p0.y
  const edge1z = v1.p0.z - v0.p0.z

  const edge2x = v2.p0.x - v0.p0.x
  const edge2y = v2.p0.y - v0.p0.y
  const edge2z = v2.p0.z - v0.p0.z

  const dUV1x = v1.u - v0.u
  const dUV1y = v1.v - v0.v
  const dUV2x = v2.u - v0.u
  const dUV2y = v2.v - v0.v

  const denom = dUV1x * dUV2y - dUV2x * dUV1y
  if (Math.abs(denom) < 1e-5) {
    const len = Math.hypot(1, 1, 1)
    return { x: 1 / len, y: 1 / len, z: 1 / len }
  }

  const f = 1.0 / denom
  return {
    x: f * (dUV2y * edge1x - dUV1y * edge2x),
    y: f * (dUV2y * edge1y - dUV1y * edge2y),
    z: f * (dUV2y * edge1z - dUV1y * edge2z),
  }
}
